metadataserver/auth/auth.go

```diff
@@ -0,0 +1,165 @@
+package auth
+
+import (
+	"crypto"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/json"
+	"strings"
+	"time"
+
+	"github.com/Netflix/titus-executor/metadataserver/identity"
+)
+
+// Authenticator generates and authenticates tokens
+type Authenticator interface {
+	GenerateToken(ttl time.Duration) (string, error)
+	VerifyToken(token string) bool
+}
+
+type envelope struct {
+	// Token token
+	Token string
+
+	// Expiration in nanoseconds
+	Expiration int64
+}
+
+// HMACAuthenticator authenticates using hmac
+type HMACAuthenticator struct {
+	Key []byte
+}
+
+// GenerateToken token
+func (a *HMACAuthenticator) GenerateToken(ttl time.Duration) (string, error) {
+	tokenBytes := make([]byte, 16)
+	_, err := rand.Read(tokenBytes)
+	if err != nil {
+		return "", err
+	}
+
+	token := base64.StdEncoding.EncodeToString(tokenBytes)
+	exp := time.Now().Add(ttl).UnixNano()
+	env := envelope{Token: token, Expiration: exp}
+	envJSON, err := json.Marshal(env)
+	if err != nil {
+		return "", err
+	}
+	envEnc := base64.StdEncoding.EncodeToString(envJSON)
+	mac := a.hmac(envEnc)
+	sig := base64.StdEncoding.EncodeToString(mac)
+
+	return envEnc + "." + sig, nil
```

Comment:
It looks like you've reimplemented / reinvented JWT / JWS here. Any reason not to just use off the shelf JWTs? They have expiration and stuff "built-in" to them?

Reply:
(I used the first implementation of Go JWT examples I found)
https://play.golang.com/p/NhQDPykRQdY


Although your (our) certificate signer implemens a slightly different interface from the one they expect. I imagine we can map it across.

Reply:
Yep, will use JWT